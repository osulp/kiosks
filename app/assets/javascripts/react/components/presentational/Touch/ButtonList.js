import {connect} from 'react-redux';
import React, {Component, PropTypes} from 'react';

export class ButtonList extends Component {
  refreshClicked() {
    this.props.fetchSlides(this.props.url);
  }
  render() {
    return (
      <li>
        <a href="#" className="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true"
           aria-expanded="false">Dropdown <span className="caret"></span></a>
        <ul className="dropdown-menu">
          <li className="refresh-slides" onClick={this.refreshClicked.bind(this)}><a>Refresh Slides</a></li>
          <li className="divider" role="separator"><a>Action</a></li>
          <li><a>Separated Link</a></li>
          <li><a>Another Link</a></li>
        </ul>
      </li>
    );
  }
}

ButtonList.propTypes = {
  url: PropTypes.string.isRequired,
  fetchSlides: PropTypes.func,
};

export default ButtonList;

