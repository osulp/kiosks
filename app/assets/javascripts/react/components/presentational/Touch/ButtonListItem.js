import {connect} from 'react-redux';
import React, {Component, PropTypes} from 'react';

export class ButtonListItem extends Component {
  onClick() {
    this.props.dispatch(this.props.onButtonClick);
  }

  render() {
    return (
      <li role={this.props.role} className={this.props.className} onClick={this.onClick.bind(this)}><a>{this.props.text}</a></li>
    );
  }
}

ButtonListItem.propTypes = {
  onButtonClick: PropTypes.func,
};

export default connect()(ButtonListItem);
