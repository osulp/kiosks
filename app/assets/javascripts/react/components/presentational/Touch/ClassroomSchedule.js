import React, {Component, PropTypes} from 'react';
import Calendar from 'rc-calendar';
import DatePicker from 'rc-calendar/lib/Picker';
import moment from 'moment';
import ConnectedClassroomScheduleDay from '../../TouchClassroomScheduleDay';
import {trackClicked} from '../shared/GoogleAnalytics';

class ClassroomSchedule extends Component {
  /**
   * Initialize a classroom schedule UI and set the initial state to display "today"
   * @param props - the properties passed into the component
   */
  constructor(props) {
    super(props);
    let now = moment();
    this.state = {selected_date: now};
  }

  /**
   * Automatically hide the classroom schedule after a period of inactivity
   */
  setHideTimeout() {
    const hide = () => {
      this.props.setModalRootComponent(undefined);
      this.props.setModalVisibility(false);
    };
    this.hide_timeout = setTimeout(hide, 120000);
  }

  /**
   * When the classroom filter element is clicked, fire off an action to propagate the change through the app state
   * @param {Object} classroom - the classroom which was clicked
   */
  classroomClicked(classroom) {
    trackClicked(this.props.google_analytics, 'TouchKiosk:ClassroomSchedule:Classroom');
    this.props.toggleClassroomSelected(classroom.shortname, classroom.selected);
  }

  /**
   * When a date on the calendar is clicked, fetch the schedule for that day and reset the hide timeout
   * @param {Moment} date - the date clicked
   */
  dateClicked(date) {
    trackClicked(this.props.google_analytics, 'TouchKiosk:ClassroomSchedule:Date');
    this.setState({selected_date: date});
    this.props.fetchClassroomSchedule(this.props.api.classroom_schedule, date);
    clearTimeout(this.hide_timeout);
    this.setHideTimeout();
  }

  /**
   * After the component has mounted, fetch the classroom list and the current days schedule
   */
  componentDidMount() {
    this.props.fetchClassrooms(this.props.api.classrooms);
    this.props.fetchClassroomSchedule(this.props.api.classroom_schedule, this.state.selected_date);
    this.setHideTimeout();
  }

  /**
   * As the component is unmounting, clear the hide timeout
   */
  componentWillUnmount() {
    clearTimeout(this.hide_timeout);
  }

  /**
   * Sort the list of classrooms by their keys
   * @returns {Array} - sorted list of classrooms
   * @private
   */
  _sortedClassrooms() {
    const ordered = [];
    Object.keys(this.props.classrooms).sort().forEach((k) => ordered.push(this.props.classrooms[k]));
    return ordered;
  }

  _hasSelectedClassroom(element, index, array) {
    return element.selected === true;
  }

  /**
   * Includes Bootstrap elements for mobile and regular displays using hidden-*  and col-* semantics
   * @returns {JSX} - the rendered UI
   */
  render() {
    const calendar = (<Calendar defaultValue={this.state.selected_date}
                                onChange={this.dateClicked.bind(this)}
                                showDateInput={false}/>);
    const is_fetching_class = this.props.is_fetching_classroom_schedule ? "is_fetching" : "";
    const date = this.state.selected_date.format('dddd, MMMM Do, YYYY');
    const classrooms = this._sortedClassrooms();
    return (
      <div id="classroom_schedule" className="panel panel-default">
        <div className="panel-heading">
          <span className="panel-title">Library Classrooms Schedule</span>
          <span className={`glyphicon glyphicon-repeat ${is_fetching_class}`} aria-hidden="true">&nbsp;</span>
        </div>
        <div className="container-fluid classroom-schedule-table-container">
          <div className="row">
            <h2>{date}</h2>
            <div className="classroom-schedule-table-datepicker hidden-md hidden-lg col-sm-12">
              <DatePicker calendar={calendar}
                          animation="slide-down"
                          defaultValue={this.state.selected_date}
                          onChange={this.dateClicked.bind(this)}
                          showDateInput={false}>
                {({value}) => {
                  return (
                    <span tabIndex="0">
                      <label htmlFor="classroom_schedule_date_picker">Select a Date:</label>
                      <input id="classroom_schedule_date_picker"
                             placeholder="Select a date."
                             style={{width: 250}}
                             readOnly
                             tabIndex="-1"
                             className="ant-calendar-picker-input ant-input"
                             value={value && value.format('YYYY-MM-DD') || ''}/>
                    </span>
                  );
                }}
              </DatePicker>
            </div>
            <div className="hidden-sm hidden-xs col-md-9 classroom-filter">
              <ul>
                <li key="filter.classroom_schedule" >Filter Schedule:</li>
                {classrooms.map((classroom) => {
                  let selected = classrooms.some(this._hasSelectedClassroom) ? classroom.selected : true;
                  return (
                    <li key={classroom.shortname}
                        className={`${classroom.shortname} ${classroom.selected ? "selected" : ""}`}
                        data-shortname={classroom.shortname}
                        data-selected={selected}
                        onClick={this.classroomClicked.bind(this, classroom)}>
                      <span className={`glyphicon glyphicon-ok-sign ${classroom.selected ? "checked" : ""}`}
                            aria-hidden="true">
                        &nbsp;
                      </span>
                      {classroom.title}
                    </li>);
                })}
              </ul>
            </div>
            <div className="col-md-9 col-sm-12">
              <ConnectedClassroomScheduleDay />
            </div>
            <div className="hidden-sm hidden-xs col-md-3">{calendar}</div>
          </div>
        </div>
      </div>
    );
  }
}

ClassroomSchedule.propTypes = {
  classrooms: PropTypes.object.isRequired,
  api: PropTypes.object.isRequired,
  google_analytics: PropTypes.func,
  is_fetching_classroom_schedule: PropTypes.bool.isRequired,
  fetchClassroomSchedule: PropTypes.func.isRequired,
  fetchClassrooms: PropTypes.func.isRequired,
  toggleClassroomSelected: PropTypes.func.isRequired,
};

export default ClassroomSchedule;
