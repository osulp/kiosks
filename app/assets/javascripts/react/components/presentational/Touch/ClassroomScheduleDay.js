import React, {Component, PropTypes} from 'react';
import moment from 'moment';

/**
 * the valid hours in which classrooms are available for
 * @type {number[]}
 */
const CLASSROOM_HOURS = [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23];

class ClassroomScheduleDay extends Component {
  /**
   * Determine if the provided time is within the start_time and end_time
   * @param {Moment} start_time - the events start time
   * @param {Moment} end_time - the events end time
   * @param {Moment} this_time - the specific time to check
   * @returns {boolean|*} - true if this_time is >= start_time and < end_time
   * @private
   */
  _eventIncludesTime(start_time, end_time, this_time) {
    // current 15 minute interval is between start and end time, `[)` momentjs
    // argument to include the start_time (this_time >= start_time) but not end_time (this_time < end_time)
    return this_time.isBetween(start_time, end_time, null, '[)');
  }

  /**
   * Display a row with the events details rendered full width.
   * @param {Object} event - the event to render details for
   * @param {number} i - the events index
   * @returns {JSX} - the TR element
   * @private
   */
  _eventDetailRow(event, i) {
    return (
      <tr key={`event.${i}.detail`} className={`event-detail-row ${event.room_shortname}`}>
        <td colSpan={CLASSROOM_HOURS.length}>
          <div>
            <span className="event-title">{event.title}</span>
            <span>{event.contact}</span>
            <span>{event.room}</span>
          </div>
        </td>
      </tr>
    )
  }

  /**
   * Given the events start_time and end_time, iterate through each of the configured classroom hours and
   * render cells for each 15min increment which the event is active. The effect is to display a cell for each hour,
   * but show it subdivided with 15min spans to give a visual representation of each interval when the event is active.
   * @param {Moment} date - the current day which the schedule is displaying
   * @param {Object} event - the event to render times for
   * @param {number} i - the events index
   * @param {willamette} true if both willamette east and west are selected in
   * the filter
   * @returns {JSX} - the TR element
   * @private
   */
  _eventTimeRow(date, event, i, willamette) {
    const start_time = moment(event.start_time);
    const end_time = moment(event.end_time);

    return (
      <tr key={`event.${i}.time`} className={`event-time-row ${event.room_shortname} ${event.combined_willamette_room? "willamette-combined" : ""} ${willamette ? "willamette-selected" : ""}`}>
        {CLASSROOM_HOURS.map((h) => {
          const hour = moment(date).hour(h);
          const has_hour = this._eventIncludesTime(start_time, end_time, hour);
          const has_15 = this._eventIncludesTime(start_time, end_time, moment(hour).add(15, 'm'));
          const has_30 = this._eventIncludesTime(start_time, end_time, moment(hour).add(30, 'm'));
          const has_45 = this._eventIncludesTime(start_time, end_time, moment(hour).add(45, 'm'));
          const hour_has_event = has_hour || has_15 || has_30 || has_45;
          return (
            <td key={`event.${i}.${hour}`} className={hour_has_event ? "event-hour" : ""}>
              <span key={`event.${i}.hour.header`} className="event-hour-header">
                {hour.format('h a')}
              </span>
              <span key={`event.${i}.00`} className={has_hour ? "event-time" : ""}>&nbsp;</span>
              <span key={`event.${i}.15`} className={has_15 ? "event-time" : ""}>&nbsp;</span>
              <span key={`event.${i}.30`} className={has_30 ? "event-time" : ""}>&nbsp;</span>
              <span key={`event.${i}.45`} className={has_45 ? "event-time" : ""}>&nbsp;</span>
            </td>
          )
        })}
      </tr>
    )
  }

  /**
   * This function is to check if both willamette east and west are selected by
   * the user in the filter section
   * @param {Object} selected - an array of rooms selected
   * @returns {bool} - true if both willamette east and west are selected
   * @private
   */
  _willametteEastWestSelected(selected) {
    const willamette_west_selected = selected.filter(function(selection){ return (selection.shortname === "lib-willwest")});
    const willamette_east_selected = selected.filter(function(selection){ return (selection.shortname === "lib-willeast")});
    return (willamette_west_selected.length > 0 && willamette_east_selected.length > 0)
  }

  /**
   * An event has a detail row followed by the event time row so that the detail can have full width of the table
   * and the time cells (15min increments) are colored and lined up properly on the x axis.
   * @param {Object} classrooms - the classrooms from the application state
   * @param {Object} events - the events from the application state
   * @param {Moment} date - the day which the schedule is being rendered
   * @returns {JSX} - rows of event details/times or indication if there are no events
   * @private
   */
  _getEventRows(classrooms, events, date) {
    if (!events || !events.length) {
      return (<tr>
        <td>No classrooms scheduled.</td>
      </tr>);
    } else {
      const rows = [];
      this._filteredEvents(classrooms, events).map((e, i) => {
        const selected = Object.values(classrooms).filter((c) => c.selected == true);
        rows.push(this._eventDetailRow(e, i));

        if (this._willametteEastWestSelected(selected) === true) {
          rows.push(this._eventTimeRow(date, e, i, true));
        } else {
          rows.push(this._eventTimeRow(date, e, i, false));
        }
 
      });
      return rows;
    }
  }


  /**
   * Show only events which relate to the currently selected classrooms from the filter
   * @param {Object} classrooms - the classrooms from the application state
   * @param {Object} events - the events from the application state
   * @returns {*} - a filtered list of events
   * @private
   */
  _filteredEvents(classrooms, events) {
    const selected = Object.values(classrooms).filter((c) => c.selected == true);
    const will_combined = [];

    if (this._willametteEastWestSelected(selected) === true) {
      // both rooms (east and west) were selected
      events.forEach(function(event_item) {
        if (event_item.combined_willamette_room === true) {
          // for combined events, exclude willamette east (keep west)
          if ((event_item.room_shortname !== "lib-willeast")) {
            will_combined.push(event_item);
          }
        } else {
          will_combined.push(event_item);
        }

      });
      return Object.values(will_combined).filter((e) => selected.some((s) => s.shortname == e.room_shortname));
    }

    if (selected.length == Object.values(classrooms).length) {
      return events;
    }
    return Object.values(events).filter((e) => selected.some((s) => s.shortname == e.room_shortname));
  }

  render() {
    const events = this.props.classroom_schedule.events;
    const classrooms = this.props.classrooms;
    let date = this.props.date;
    if (date == '') {
      return null;
    }
    date = moment(date);
    return (
      <table className="table classroom-schedule-table">
        <thead>
        <tr key="event.header" className="event-header-row">
          <th colSpan={CLASSROOM_HOURS.length}>
            <div>
              <span>Event Title</span>
              <span>Contact</span>
              <span>Location</span>
            </div>
          </th>
        </tr>
        </thead>
        <tbody>
        {this._getEventRows(classrooms, events, date)}
        </tbody>
      </table>);
  }
}

ClassroomScheduleDay.propTypes = {
  date: PropTypes.string.isRequired,
  classrooms: PropTypes.object.isRequired,
  classroom_schedule: PropTypes.object.isRequired,
  is_fetching_classroom_schedule: PropTypes.bool.isRequired,
};

export default ClassroomScheduleDay;
