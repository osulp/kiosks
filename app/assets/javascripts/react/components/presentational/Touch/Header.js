import React, {Component, PropTypes} from 'react';
import TouchHours from '../../TouchHours';
import ConnectedTabbedPanel from '../../TabbedPanel';
import ConnectedMaps from '../../TouchMaps';
import ConnectedSearchPrimo from '../../TouchSearchPrimo';
import {trackClicked} from '../shared/GoogleAnalytics';

class Header extends Component {
  constructor(props) {
    super(props);
    this.state = { taps: 0 };
  }

  /**
   * Increase the state taps counter set a timer to reset it in 2 seconds unless another tap event happens. This
   * matters to the view in so much as it's able to determine if there have been continuous taps without 2 seconds of
   * delay in between. It's a feature..
   * @private
   */
  _didTap() {
    trackClicked(this.props.google_analytics, 'TouchKiosk:Header:Logo');
    this.setState({taps: this.state.taps + 1});
    clearTimeout(this.reset_timeout);
    const reset = () => {
      this.setState({taps: 0});
    };
    this.reset_timeout = setTimeout(reset, 2000);
  }

  /**
   * Refresh button was clicked, fire off actions to fetch the latest slides and reset
   * the gallery scroll position
   */
  refreshClicked() {
    trackClicked(this.props.google_analytics, 'TouchKiosk:Header:Refresh');
    this.props.fetchSlides(this.props.url);
    this.props.scrollToSlide(0);
  }

  /**
   * Hours button was clicked, set the modal to display the connected Hours component
   */
  hoursClicked() {
    trackClicked(this.props.google_analytics, 'TouchKiosk:Header:Hours');
    this.props.setModalVisibility(true);
    this.props.setModalRootComponent(<TouchHours />);
  }

  /**
   * Maps button was clicked, set the modal to display the connected tab panel component
   */
  mapsClicked() {
    trackClicked(this.props.google_analytics, 'TouchKiosk:Header:Maps');
    let tabs = this.props.maps.map((m) => {
      return {
        button_text: m.title,
        content: (<div style={{backgroundImage: `url("${m.image_url}")`}}>&nbsp;</div>)
      };
    });
    this.props.setModalVisibility(true);
    this.props.setModalRootComponent(<ConnectedTabbedPanel id="maps_tabbed_panel"
                                                           tabs={tabs}
                                                           selectedIndex={1}
                                                           timeout={30000}/>);
  }

  /**
   * The 1search button was clicked, set the modal to display the connected search primo component
   */
  searchPrimoClicked() {
    trackClicked(this.props.google_analytics, 'TouchKiosk:Header:SearchPrimo');
    this.props.setModalVisibility(true);
    this.props.setModalRootComponent(<ConnectedSearchPrimo />);
  }
  /**
   * The maps button was clicked, set the modal to display the maps component
   */
  mapsClicked() {
    trackClicked(this.props.google_analytics, 'TouchKiosk:Header:Maps');
    this.props.setModalVisibility(true);
    this.props.setModalRootComponent(<ConnectedMaps />);
  }

  /**
   * Format text based on the hours that were fetched for "today", defaulting to an empty string if there was an error or
   * no hours were fetched due to issues on the servers dependency to the data source.
   * @returns {String} - the hours detail
   * @private
   */
  _hoursToday(){
    if(this.props.hours.error || Object.values(this.props.hours).length == 0) {
      return "";
    } else {
      let today = Object.values(this.props.hours)[0];
      return `${today.formatted_hours.trim()} on ${today.string_date}`;
    }
  }

  render() {
    let tapped_enough = (this.state.taps > 20);
    let is_fetching = this.props.is_fetching_slides ? "is_fetching" : "";
    return (
      <div className="navbar-wrapper">
        <div className="container">
          <nav className="navbar navbar-inverse navbar-static-top">
            <div className="container">
              <div className="navbar-header">
                <button type="button" className="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar"
                        aria-expanded="false" aria-controls="navbar">
                  <span className="sr-only">Toggle navigation</span>
                  <span className="icon-bar"></span>
                  <span className="icon-bar"></span>
                  <span className="icon-bar"></span>
                </button>
                <img className={`logo ${tapped_enough ? 'is_active' : ''}`} src="/images/osulibrarylogo.png" onClick={this._didTap.bind(this)} />
              </div>
              <div id="navbar" className="navbar-collapse collapse">
                <ul className="nav navbar-nav">
                  <li className="show-hours" onClick={this.hoursClicked.bind(this)}>
                    <a className="btn btn-navbar btn-default">Hours</a>
                  </li>
                  <li className="show-maps" onClick={this.mapsClicked.bind(this)}>
                    <a className="btn btn-navbar btn-default">Maps</a>
                  </li>
                  <li className="show-search-primo" onClick={this.searchPrimoClicked.bind(this)}>
                    <a className="btn btn-navbar btn-default">1Search</a>
                  </li>
                </ul>
                <ul className="nav navbar-nav navbar-right">
                  <li className="refresh-slides" onClick={this.refreshClicked.bind(this)}>
                    <a>
                      <span
                        className={`glyphicon glyphicon-repeat ${is_fetching}`}
                        aria-hidden="true">&nbsp;</span>
                    </a>
                  </li>
                </ul>
                <p className="hours navbar-text">{this._hoursToday()}</p>
              </div>
            </div>
          </nav>
        </div>
      </div>
    );
  }
}

Header.propTypes = {
  url: PropTypes.string.isRequired,
  maps: PropTypes.array,
  hours: PropTypes.object.isRequired,
  google_analytics: PropTypes.func,
  is_fetching_slides: PropTypes.bool.isRequired,
  fetchSlides: PropTypes.func.isRequired,
  setModalVisibility: PropTypes.func.isRequired,
  setModalRootComponent: PropTypes.func.isRequired,
  scrollToSlide: PropTypes.func.isRequired
};

export default Header;
