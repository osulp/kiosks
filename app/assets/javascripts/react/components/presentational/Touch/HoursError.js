import React, {Component, PropTypes} from 'react';

class HoursError extends Component {
  render() {
    return (<p className="hours-table-error">
      <span className="glyphicon glyphicon-exclamation-sign"> </span>
      <span>{this.props.error}</span>
    </p>);
  }
}

HoursError.propTypes = {
  error: PropTypes.string.isRequired,
};

export default HoursError;
