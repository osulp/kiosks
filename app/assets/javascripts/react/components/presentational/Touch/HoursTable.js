import React, {Component, PropTypes} from 'react';
import moment from 'moment';

class HoursTable extends Component {
  render() {
    let hours = this.props.hours;
    return (
      <table className="table hours-table">
        <thead>
        <tr>
          <th></th>
          <th>Hours</th>
        </tr>
        </thead>
        <tbody>
        {Object.values(hours).map((h, i) => {
          return (
            <tr key={`hours.${i}`} className={moment(h.sortable_date).isSame(this.props.selected_date, "day") ? "alert alert-info" : ""}>
              <td>{h.string_date}</td>
              <td>{h.formatted_hours}</td>
              <td>{h.event_desc}</td>
            </tr>
          )
        })}
        </tbody>
      </table>);
  }
}

HoursTable.propTypes = {
  hours: PropTypes.object.isRequired,
  selected_date: PropTypes.object.isRequired,
};

export default HoursTable;
