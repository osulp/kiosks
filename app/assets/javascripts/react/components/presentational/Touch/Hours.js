import React, {Component, PropTypes} from 'react';
import Calendar from 'rc-calendar';
import DatePicker from 'rc-calendar/lib/Picker';
import moment from 'moment';
import HoursTable from './HoursTable';
import HoursError from './HoursError';
import {trackClicked} from '../shared/GoogleAnalytics';

export const getWeekArray = (date) => {
  return ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"].map(d => moment(date).day(d).format('YYYY-MM-DD'));
};

class Hours extends Component {
  /**
   * Initialize an hours UI and set the initial state to display "today"
   * @param props - the properties passed into the component
   */
  constructor(props) {
    super(props);
    let now = moment();
    this.state = {selected_date: now};
  }

  /**
   * Set the modal to automatically hide itself after a period of time, unless the timeout is cleared beforehand
   */
  setHideTimeout() {
    this.hide_timeout = setInterval(() => {
      this.props.setModalRootComponent(undefined);
      this.props.setModalVisibility(false);
    }, 2*10*1000);
  }

  /**
   * Change the component to display the date that was clicked, and reset the hide timeout.
   * @param {Moment} date - the date that was clicked
   */
  dateClicked(date) {
    trackClicked(this.props.google_analytics, 'TouchKiosk:Hours:Date');
    this.setState({selected_date: date});
    this.props.fetchHours(this.props.api.hours, getWeekArray(date));
    clearInterval(this.hide_timeout);
    this.setHideTimeout();
  }

  /**
   * After the component mounts, fetch a weeks worth of hours for the date that was selected (defaulting to 'now')
   */
  componentDidMount() {
    this.setHideTimeout();
    this.props.fetchHours(this.props.api.hours, getWeekArray(moment().format('YYYY-MM-DD')));
  }

  /**
   * Before the component unmounts, clear the hide timeout
   */
  componentWillUnmount() {
    this.props.fetchHours(this.props.api.hours, [moment().format('YYYY-MM-DD')]);
    clearInterval(this.hide_timeout);
  }

  /**
   * Render the hours table or error depending on the result of fetching the hours
   * @param {Object} hours - the hours returned from the server
   * @returns {JSX} - the HoursTable or HoursError depending on the servers return
   * @private
   */
  _hoursContent(hours) {
    if (hours.error) {
      return (<HoursError error={hours.error}/>);
    } else {
        if(typeof this.state.selected_date == 'string') {
            return (<HoursTable hours={hours} selected_date={moment(this.state.selected_date)}/>);
        } else {
            return (<HoursTable hours={hours} selected_date={this.state.selected_date}/>);
        }
    }
  }

  /**
   * Includes Bootstrap elements for mobile and regular displays using hidden-*  and col-* semantics
   * @returns {JSX} - the rendered UI
   */
  render() {
    let hours = this.props.hours;
    let calendar = (<Calendar defaultValue={this.state.selected_date}
                              onChange={this.dateClicked.bind(this)}
                              showDateInput={false}/>);
    let is_fetching_class = this.props.is_fetching_hours ? "is_fetching" : "";
    return (
      <div id="hours" className="panel panel-default">
        <div className="panel-heading">
          <span className="panel-title">Library Hours</span>
          <span className={`glyphicon glyphicon-repeat ${is_fetching_class}`} aria-hidden="true">&nbsp;</span>
        </div>
        <div className="container-fluid hours-table-container">
          <div className="row">
            <div className="hours-table-datepicker hidden-md hidden-lg col-sm-12">
              <DatePicker calendar={calendar}
                          animation="slide-down"
                          defaultValue={this.state.selected_date}
                          onChange={this.dateClicked.bind(this)}
                          showDateInput={false}>
                {({value}) => {
                  return (
                    <span tabIndex="0">
                      <label htmlFor="hours_date_picker">Select a Date:</label>
                      <input id="hours_date_picker"
                             placeholder="Select a date."
                             style={{width: 250}}
                             readOnly
                             tabIndex="-1"
                             className="ant-calendar-picker-input ant-input"
                             value={value && value.format('YYYY-MM-DD') || ''}/>
                    </span>
                  );
                }}
              </DatePicker>
            </div>
            <div className="col-md-5 col-sm-12">{this._hoursContent(hours)}</div>
            <div className="hidden-sm hidden-xs col-md-7">{calendar}</div>
          </div>
        </div>
      </div>
    );
  }
}

Hours.propTypes = {
  hours: PropTypes.object.isRequired,
  api: PropTypes.object.isRequired,
  google_analytics: PropTypes.func,
  is_fetching_hours: PropTypes.bool.isRequired,
  fetchHours: PropTypes.func.isRequired,
};

export default Hours;
