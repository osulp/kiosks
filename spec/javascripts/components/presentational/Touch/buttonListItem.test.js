import React from 'react';
import { ButtonListItem } from '../../../../../app/assets/javascripts/react/components/presentational/Touch/ButtonListItem';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

describe('Snapshot', () => {

  const props = {
    role: "main",
    className: "class",
    text: "blah",
    onButtonClick: jest.fn()
  };

  const component = renderer.create(
    <ButtonListItem {...props} />
  );

  it('should match the cached snapshot', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('runs a callback onClick', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    tree.props.onClick();
    expect(props.onButtonClick.mock.calls.length).toEqual(1);
  });

  it('does not change the state of the component onClick', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    tree.props.onClick();

    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
