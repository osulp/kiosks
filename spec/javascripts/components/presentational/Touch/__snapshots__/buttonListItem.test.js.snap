exports[`Snapshot does not change the state of the component onClick 1`] = `
<li
  className="class"
  onClick={[Function]}
  role="main">
  <a>
    blah
  </a>
</li>
`;

exports[`Snapshot does not change the state of the component onClick 2`] = `
<li
  className="class"
  onClick={[Function]}
  role="main">
  <a>
    blah
  </a>
</li>
`;

exports[`Snapshot runs a callback onClick 1`] = `
<li
  className="class"
  onClick={[Function]}
  role="main">
  <a>
    blah
  </a>
</li>
`;

exports[`Snapshot should match the cached snapshot 1`] = `
<li
  className="class"
  onClick={[Function]}
  role="main">
  <a>
    blah
  </a>
</li>
`;
